utils/sub8_ros_tools/so3tools/so3tools.py

```diff
@@ -0,0 +1,557 @@
+    True
+
+    """
+    # sorry Mario, but the function you are looking for is in another castle
+    angle, axis = angle_axis(x)
+    return angle * axis
+
+
+def angle_axis(x):
+    """Returns the extracted angle and axis from a valid SO3 quantity x.
+    This is equivalent to separating the axle form of x into its
+    magnitude and direction. The angle will always be between 0 and pi,
+    inclusive, and the axis will always be a unit vector.
+
+    >>> yourAngle, yourAxis = -1337, [-2,0,7]
+    >>> yourMat = trns.rotation_matrix(yourAngle, yourAxis)
+    >>> myAngle, myAxis = angle_axis(yourMat)
+    >>> print(np.isclose(myAngle, np.mod(yourAngle, 2*np.pi)))
+    True
+    >>> print(np.allclose(myAxis, trns.unit_vector(yourAxis)))
+    True
+    >>> myAngle2, myAxis2 = angle_axis(trns.quaternion_from_matrix(yourMat))
+    >>> print(np.allclose(myAngle, myAngle2), np.allclose(myAxis, myAxis2))
+    (True, True)
+
+    """
+    xrep = rep(x)
+    # In the matrix case, transformations.py already has a nice implementation:
+    if xrep in ['rotmat', 'tfmat']:
+        if xrep == 'rotmat':
+            x = affine(x)
+        angle, axis, point = trns.rotation_from_matrix(x)
+        # But to be consistent with axles, we only use positive angles:
+        if angle < 0:
+            angle, axis = -angle, -axis
+        # And we map the identity rotation to an axis of [0,0,0]:
+        elif np.isclose(angle, 0):
+            axis = np.array([0, 0, 0])
+    # In the quaternion case, we carry out the following routine:
+    elif xrep == 'quaternion':
+        # Renormalize for accuracy:
+        x = trns.unit_vector(x)
+        # If "amount of rotation" is negative, flip quaternion:
+        if x[0] < 0:
+            x = -x
+        # Extract axis:
+        imMax = max(map(abs, x[1:]))
+        if not np.isclose(imMax, 0):
+            axis = trns.unit_vector(x[1:] / (imMax * npl.norm(x[1:])))
+        else:
+            axis = np.array([0, 0, 0])
+        # Extract angle:
+        angle = 2 * np.arccos(x[0])
+    # In the axle case, the routine is trivial:
+    elif xrep == 'axle':
+        angle = npl.norm(x)
+        if not np.isclose(angle, 0):
+            axis = x / angle
+        else:
+            axis = np.array([0, 0, 0])
+    # If not SO3:
+    else:
+        raise ValueError('Quantity to extract angle and axis from must be on SO3')
+    # Finally:
+    return (angle, axis)
+
+
+def quaternion_from_axle(r):
+    """Returns the quaternion equivalent to the given axle r.
+
+    >>> myAxle = np.pi * np.array([0,1,0])
+    >>> q = quaternion_from_axle(myAxle)
+    >>> M1 = trns.quaternion_matrix(q)
+    >>> M2 = trns.rotation_matrix(np.pi, [0,1,0])
+    >>> print(trns.is_same_transform(M1, M2))
+    True
+
+    """
+    angle = np.mod(npl.norm(r), 2 * np.pi)
+    if not np.isclose(angle, 0):
+        return trns.quaternion_about_axis(angle, r / angle)
+    else:
+        return np.array([1, 0, 0, 0])  # unit real number is identity quaternion
```

Comment:
Correct me if I'm wrong, but I think that the convention that transformations.py uses is XYZW, so this isn't the identity quaternion. At least on my system:

    >>> from tf import transformations
    >>> transformations.quaternion_about_axis(0, [1, 0, 0])
    array([ 0.,  0.,  0.,  1.])